testsuite/cluster-test/src/deployment.rs

```diff
@@ -117,19 +126,61 @@ impl DeploymentManager {
     }
 
     pub fn tag_tested_image(&mut self, hash: String) -> failure::Result<()> {
-        self.tag_image(TESTED_TAG.to_string(), hash.clone())?;
+        let image_id = ImageIdentifier {
+            image_digest: Some(hash.clone()),
+            image_tag: None,
+        };
+        self.tag_image(VALIDATOR_IMAGE_NAME, &image_id, TESTED_TAG)?;
+        let upstream_tag = self.get_upstream_tag(&image_id)?;
+        self.tag_image(
+            CLIENT_IMAGE_NAME,
+            &ImageIdentifier {
+                image_digest: None,
+                image_tag: Some(upstream_tag.clone()),
+            },
+            TESTED_TAG,
+        )?;
+        self.tag_image(
+            FAUCET_IMAGE_NAME,
+            &ImageIdentifier {
+                image_digest: None,
+                image_tag: Some(upstream_tag),
+            },
+            TESTED_TAG,
+        )?;
+
         fs::write(LAST_DEPLOYED_FILE, &hash).expect("Failed to write .last_deployed_digest");
         self.last_deployed_digest = Some(hash);
         Ok(())
     }
 
-    fn tag_image(&self, tag: String, hash: String) -> failure::Result<()> {
+    fn get_upstream_tag(&self, image_id: &ImageIdentifier) -> failure::Result<String> {
+        let images = self.get_images(VALIDATOR_IMAGE_NAME, image_id)?;
+        let default_image = &ImageIdentifier::default();
+        let default_tag = &"".to_string();
+        for image in images {
+            let tag = image
```

Comment:
I think what you want to do is something like that:
```
let image_id = match image. image_id {
   Some(image_id) => image_id;
   None => continue;
};
let image_tag = match  image_id.image_tag {} {...
```

Reply:
Updated